File: 09-HASHING/Go/design-hashmap.go
package main

// https://leetcode.com/problems/design-hashmap/

type Node struct {
	key   int
	value int
	next  *Node
}

type MyHashMap struct {
	set [1000003]*Node
}

func Constructor() MyHashMap {
	return MyHashMap{}
}

func (this *MyHashMap) Put(key, value int) {
	tail, curr := this.getNode(key)
	if curr != nil {
		curr.value = value
		return
	}

	node := &Node{key: key, value: value}
	if tail == nil {
		this.set[this.index(key)] = node
	} else {
		tail.next = node
	}
}

func (this *MyHashMap) Remove(key int) {
	prev, curr := this.getNode(key)
	if curr == nil {
		return
	}

	if prev == nil {
		this.set[this.index(key)] = curr.next
	} else {
		prev.next = curr.next
	}
}

func (this *MyHashMap) Get(key int) int {
	_, curr := this.getNode(key)
	if curr == nil {
		return -1
	}

	return curr.value
}

func (this *MyHashMap) getNode(key int) (*Node, *Node) {
	var prev *Node
	curr := this.set[this.index(key)]
	for curr != nil {
		if curr.key == key {
			return prev, curr
		}
		prev = curr
		curr = curr.next
	}
	return prev, nil
}

func (this *MyHashMap) index(key int) int {
	return key % 1000003
}

File: 09-HASHING/Python/design-hashmap.py
# https://leetcode.com/problems/design-hashmap/

class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.next = None


class MyHashMap:
    size = 1_000_003

    def __init__(self):
        self.set = [None] * self.size

    def put(self, key: int, value: int) -> None:
        tail, curr = self._get(key)
        if curr is not None:
            curr.value = value
            return

        node = Node(key=key, value=value)
        if tail is None:
            self.set[self._index(key)] = node
        else:
            tail.next = node

    def remove(self, key: int) -> None:
        prev, curr = self._get(key)
        if curr is None:
            return

        if prev is None:
            self.set[self._index(key)] = curr.next
        else:
            prev.next = curr.next

    def get(self, key: int) -> int:
        node = self._get(key)[1]
        return -1 if node is None else node.value

    def _index(self, key: int) -> int:
        return key % self.size

    def _get(self, key: int):
        index = self._index(key)
        curr = self.set[index]
        prev = None
        while curr:
            if curr.key == key:
                return prev, curr
            prev = curr
            curr = curr.next
        return prev, None
